File: 高级数据结构与算法/Python/第3章 排序算法/3.3 归并排序/merge_sort.py
def binary_insertion_sort(lst, left, right):
    for i in range(left+1, right+1):
        temp = lst[i]
        start = left
        end = i - 1
        while start <= end:
            mid = start + (end - start) // 2
            if lst[mid] > temp:
                end = mid - 1
            else:
                start = mid + 1
        j = i - 1
        while j > end:
            lst[j+1] = lst[j]
            j -= 1
        lst[j+1] = temp

def merge(lst, start, mid, end, temp):
    i = start
    j = mid + 1
    k = 0

    while i <= mid and j <= end:
        if lst[i] <= lst[j]:
            temp[k] = lst[i]
            k += 1
            i += 1
        else:
            temp[k] = lst[j]
            k += 1
            j += 1

    while i <= mid:
        temp[k] = lst[i]
        k += 1
        i += 1
    while j <= end:
        temp[k] = lst[j]
        k += 1
        j += 1

    for i in range(k):
        lst[start+i] = temp[i]

def merge_sort_worker(lst, start, end, temp):
    # 列表长度小于10时，采用二分插入排序
    if end - start <= 10:
        binary_insertion_sort(lst, start, end)
        return
    if start < end:
        mid = start + (end - start) // 2
        merge_sort_worker(lst, start, mid, temp)
        merge_sort_worker(lst, mid+1, end, temp)
        # 避免不必要的合并
        if lst[mid] <= lst[mid+1]:
            return
        merge(lst, start, mid, end, temp)

def merge_sort(lst):
    n = len(lst)
    temp = [0] * n      # 临时列表
    merge_sort_worker(lst, 0, n-1, temp)

def main():
    lst = [5, 8, 6, 3, 9, 2, 1, 7]
    merge_sort(lst)
    print(lst)

if __name__ == "__main__":
    main()

File: code/C/第13章 排序算法/13.2 归并排序/merge_sort.c
#include <stdio.h>

void showArr(int *arr, int n) {
    for(int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

void binaryInsertionSort(int *arr, int left, int right) {
    for(int i = left + 1; i <= right; i++) {
        int temp = arr[i];
        int start = left;
        int end = i - 1;
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(arr[mid] > temp) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        int j;
        for(j = i - 1; j > end; j--) {
            arr[j+1] = arr[j];
        }
        arr[j+1] = temp;
    }
}

void merge(int *arr, int start, int mid, int end, int *temp) {
    int i = start;
    int j = mid + 1;
    int k = 0;

    while(i <= mid && j <= end) {
        if(arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }

    while(i <= mid) {
        temp[k++] = arr[i++];
    }
    while(j <= end) {
        temp[k++] = arr[j++];
    }

    for(int i = 0; i < k; i++) {
        arr[start+i] = temp[i];
    }
}

void mergeSortWorker(int *arr, int start, int end, int *temp) {
    // 列表长度小于10时，采用二分插入排序
    if(end - start <= 10) {
        binaryInsertionSort(arr, start, end);
        return;
    }
    if(start < end) {
        int mid = start + (end - start) / 2;
        mergeSortWorker(arr, start, mid, temp);
        mergeSortWorker(arr, mid+1, end, temp);
        // 避免不必要的合并
        if(arr[mid] <= arr[mid+1]) {
            return;
        }
        merge(arr, start, mid, end, temp);
    }
}

void mergeSort(int *arr, int n) {
    int temp[n];        // 临时数组
    mergeSortWorker(arr, 0, n-1, temp);
}

int main() {
    int arr[] = {5, 8, 6, 3, 9, 2, 1, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    mergeSort(arr, n);
    showArr(arr, n);
    return 0;
}